src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp

```diff
@@ -139,6 +139,7 @@ smt2_incremental_decision_proceduret::smt2_incremental_decision_proceduret(
     smt_set_option_commandt{smt_option_produce_modelst{true}});
   solver_process->send(smt_set_logic_commandt{
     smt_logic_quantifier_free_uninterpreted_functions_bit_vectorst{}});
+  solver_process->send(object_size_function.declaration);
```

Comment:
General question that popped to mind:

Does it make sense to decouple the generation of any SMT terms/functions from the sending them to the solver process into a two step process?

I believe the current design has some space for any failure in any of these generation functions to become a pain in debugging because we might be sending empty strings to the solver. Or am I mistaken? I'm not very familiar with the failure modes of these functions.

Reply:
The generation of the SMT commands/terms/functions should already be sufficiently decoupled from sending them to the solver. The line in particular on which you have commented is only sending the existing function declaration data structure for conversion to string and then to the solver. It does not carry out the construction or the conversion itself. The implementations of the data structures and the conversions are in separate files and are separately unit tested. For clarity (and the basis of future documentation) the complete data flow is -

1. `exprt` - Rest of CBMC calls the decision procedure, passing various expressions. See the doxygen for `set_to` and `handle` in `src/solvers/decision_procedure.h`
2. `smt_commandt` - `smt2_incremental_decision_proceduret` uses `convert_expr_to_smt` along with the constructors of the various sub classes to construct the smt commands.
3. `std::string` - The decision procedure sends the commands to a solver process, which may be mocked for unit tests. When not mocking, `smt_piped_solver_processt::send` will use `smt_to_smt2_string` to convert the commands to strings.
4. SMT Solver - The `smt_piped_solver_processt` sends the string to the instance of `piped_processt` which manages the actual solver process.
5. In the case where the rest of CBMC requests satisfiability status / or values from the the decision procedure via the `dec_solve` or `get` member functions, the response from the solver is fetched via `smt_piped_solver_processt::receive_response`
6. `std::string` response - `smt_piped_solver_processt::receive_response` is called, which fetches the response as a string back from the solver.
7. `smt_responset` - The `recieve_response` function uses `smt2irep` and `validate_smt_response` to convert and validate the string response from the solver to produce the `smt_reponset`.
8. `decision_proceduret::result` / `exprt` - `smt2_incremental_decision_proceduret` uses the `smt_responset` to either produce a result or expression as appropriate. It the case of expressions, this is done via the separate `construct_value_expr_from_smt` function.

In regards to the possibility of sending the solver empty strings - `solver_process->send` takes an `smt_commandt` which is effectively non-nullable. It has the default constructor deleted and the `irept` constructors inherited as protected. So the only the way to construct one is by constructing one of the derived classes and upcasting. The conversion to string is intended to full fill a requirement that all commands are convertible to non-empty strings. Therefore any conditions which would otherwise result in sending an empty string should actually cause one of -

- Compile error due to an invalid attempt to empty/null construct the command/term/sort.
- Invariant violation in construction of `smt_commandt` sub-class.
- Invariant violation in conversion from `smt_commandt` to string.

The above outlined design should make it highly unlikely for an empty string to be sent to the solver. The strings sent including any empty strings can be seen in the debug level logging. We could (trivially) make certain that we don't send empty strings, by adding a `CHECK_RETURN` invariant in `smt_piped_solver_processt::send` as all conversion and sending is dispatched through this member function.

The conditions which might be tricky to debug and could be worth considering improving our error handling on, in my opinion would be -

1. Starting of the solver process - See constructor of `smt_piped_solver_processt` and `solver_factoryt::get_incremental_smt2`. This could do with some better checking that the user specified solver command actually starts a solver and has not returned an error code or otherwise terminated prematurely.
2. Non return of the solver. This is the case where the solver is either solving indefinitely or moves into a state of waiting for further input instead of returning a response when we expect a response. This would be trickier to improve as we would need to come up with a way to distinguish long processing times from error states.

If you have further questions about these issues then it might be more efficient to discuss them outside of github review comments.